torch/distributed/_spmd/api.py

```diff
@@ -429,22 +445,59 @@ def stateless_func(func, params, buffers, named_states, args, kwargs):
         ), _rematerialize_optimizer(
             opt, named_states, params
         ) if opt else nullcontext():
+            # For DataParallel mode, install hooks first to tag the graidents
+            # we would remove those nodes later, this is safe to trace
+            tagging_hooks = []
+            if isinstance(parallel_mode, DataParallel):
+                for p in params.values():
+                    h = p.register_hook(lambda grad: torch.ops._spmd.tag_grad(grad))
+                    tagging_hooks.append(h)
+
             ret = func(*args, **kwargs)
+
+            # clear up the hooks after the train step
+            for h in tagging_hooks:
+                h.remove()
+
             # make sure updated parameters are returned
             return ret, list(mod.parameters()), list(named_states.values())  # type: ignore[union-attr]
 
-    # FIXME: Using symbolic tracing to work around. Otherwise it hits
-    # shape mismatch error, as we use local inputs to trace local graph
-    # and use DTensor to expand operators, where DTensor's shape is the
-    # global shape.
+    is_data_parallel_mode = isinstance(parallel_mode, DataParallel)
+
+    # FIXME: Using symbolic tracing to work around in DTensor expand mode.
+    # Otherwise it hits shape mismatch error, as we use local inputs to
+    # trace local graph and use DTensor to expand operators, where
+    # DTensor's shape is the global shape.
+    tracing_mode = "fake" if is_data_parallel_mode else "symbolic"
```

Comment:
there is a distinction to draw here.

Using symbolic tracing mode helps us produce safe traces, that's all.  We start with real inputs, and if we don't call any strange ops like nonzero we won't get any symints along the way either.  we still get to do inductor optimizations relative to the parts of the trace that are real, which in your case with these typical models, is probably the whole model.  The upside of using symbolic tracing is, if some op in a model does produce a symint, we safely trace that instead of producing an unsafe trace and not knowing it.

Enabling dynamic shapes in PT2 is different.  Its first version started with all input dims being symints, and producing a hard-to-optimize program.  The most recent incarnation does a trick called auto dynamism, where it actually traces/compiles using static input shapes on the first try, then notices by observing which guards failed on iter2, and retraces with only those input dims as dynamic.

Reply:
Yeah I think it depends on whether the PT2 want to enable dynamic shape by default in the future as that would be a safe trace as @wconstab mentioned, we don't really need to buy in symbolic shape at this point. 

About what I am doing in this PR, I want to clarify: This PR does not use symbolic tracing, the previous DTensorExpand approach used symbolic tracing to deal with global shape/local shape mismatch issue. This PR just tries to preserve the old behavior and it uses "fake" mode tracing instead which produce static shapes.